lib/segment/src/index/field_index/geo_index.rs

```diff
@@ -1 +1,533 @@
+    }
+
+    /// number of geohashes within the rectangle
+    fn area_region_count(&self) -> usize {
+        self.area_height_region_count() * self.area_width_region_count()
+    }
+
+    fn compute_from_bounding_box(bounding_box: &GeoBoundingBox, geohash_precision: usize) -> Self {
+        let GeoPoint {
+            lat: max_lat,
+            lon: min_lon,
+        } = bounding_box.top_left;
+        let GeoPoint {
+            lat: min_lat,
+            lon: max_lon,
+        } = bounding_box.bottom_right;
+
+        // Unwrap is acceptable, as data should be validated before
+        let north_west = encode((min_lon, max_lat).into(), geohash_precision).unwrap();
+        let south_west = encode((min_lon, min_lat).into(), geohash_precision).unwrap();
+        let south_east = encode((max_lon, min_lat).into(), geohash_precision).unwrap();
+        let north_east = encode((max_lon, max_lat).into(), geohash_precision).unwrap();
+
+        Self {
+            north_west,
+            south_west,
+            south_east,
+            north_east,
+        }
+    }
+}
+
+/// Return as-high-as-possible with maximum of `max_regions`
+/// number of geo-hash guaranteed to contain the whole circle.
+fn circle_hashes(circle: GeoRadius, max_regions: usize) -> Vec<GeoHash> {
+    let rectangle = minimum_bounding_rectangle_for_circle(&circle);
+    let inner_hashes = rectangle_hashes(rectangle, max_regions);
+    filter_hashes_within_circle(inner_hashes, circle)
+}
+
+/// filter geo hashes for which the center is within a circle
+fn filter_hashes_within_circle(hashes: Vec<GeoHash>, circle: GeoRadius) -> Vec<GeoHash> {
+    let radius_km = circle.radius / 1000.0;
+    hashes
+        .into_iter()
+        .map(|h| {
+            let coord = decode(&h).unwrap().0;
+            let geo_point = GeoPoint {
+                lon: coord.x,
+                lat: coord.y,
+            };
+            (h, geo_point)
+        })
+        .filter(|(_, hash_point)| {
+            distance_in_km_between_coordinates(&circle.center, hash_point) <= radius_km
+        })
+        .map(|(h, _)| h)
+        .collect()
+}
+
+/// Return as-high-as-possible with maximum of `max_regions`
+/// number of geo-hash guaranteed to contain the whole rectangle.
+fn rectangle_hashes(rectangle: GeoBoundingBox, max_regions: usize) -> Vec<GeoHash> {
+    let full_geo_bounding_box: FullGeoBoundingBox = (&rectangle).into();
+    let shortest_side_km = full_geo_bounding_box.shortest_side_length_in_km();
+    let possible_precisions = geohash_precisions_for_distance(shortest_side_km * 1000.0);
+    // filter precision which generates less than `max_regions`.
+    let rectangle_geohash = possible_precisions
+        .into_iter()
+        .map(|p| RectangleGeoHash::compute_from_bounding_box(&rectangle, p))
+        .take_while(|rectangle_geohash| rectangle_geohash.area_region_count() <= max_regions)
+        .last();
+    match rectangle_geohash {
+        None => Vec::new(),
```

Comment:
> in what cases it could be empty? I guess one reason is if max_regions = 0?

In my test cases it happens rather often.

Currently the candidates geohash precisions have dimensions which are smaller than the shorter side of the rectangle.
This way we get to have several geohashes to encode the total rectangle area. This increase the granularity.

But this also makes the algorithm more restrictive and excludes the case where `max_regions = 1` by design.

We could include in the candidate geohash precisions, the smallest precision which wraps completely the area, this way we would always cover the case where `max_regions = 1` and yield at least one result. (there is a special case, this does not work if the area is bigger than a single top geohash regions).

WDYT?

Reply:
> candidates geohash precisions have dimensions which are smaller than the shorter side of the rectangle

that does not guarantee that the tile of this precision will include the whole rectangle, does it?

> But this also makes the algorithm more restrictive and excludes the case where max_regions = 1 by design.

It should include

> there is a special case, this does not work if the area is bigger than a single top geohash regions

That should result in empty geo-hash `""`

Reply:
I tried to capture your comment in the following commit https://github.com/qdrant/qdrant/pull/315/commits/5cb11072465a04cb355f3a8df636f84ff7de068d

It makes sure that no empty vector.